include/multiphase/private/fd_operators_inc.h

```diff
@@ -0,0 +1,514 @@
+                               ((*thn_data)(idx_c_up) * ((*un_data)(idx + xp) - (*un_data)(idx)) -
+                                (*thn_data)(idx_c_low) * ((*un_data)(idx) - (*un_data)(idx - xp)));
+        double ddy_Thn_dy_un = eta_n / (dx[1] * dx[1]) *
+                               (thn_imhalf_jphalf * ((*un_data)(idx + yp) - (*un_data)(idx)) -
+                                thn_imhalf_jmhalf * ((*un_data)(idx) - (*un_data)(idx - yp)));
+        double ddy_Thn_dx_vn = eta_n / (dx[1] * dx[0]) *
+                               (thn_imhalf_jphalf * ((*un_data)(upper_y_idx) - (*un_data)(u_y_idx)) -
+                                thn_imhalf_jmhalf * ((*un_data)(lower_y_idx) - (*un_data)(l_y_idx)));
+        double ddx_Thn_dy_vn = -eta_n / (dx[0] * dx[1]) *
+                               ((*thn_data)(idx_c_up) * ((*un_data)(upper_y_idx) - (*un_data)(lower_y_idx)) -
+                                (*thn_data)(idx_c_low) * ((*un_data)(u_y_idx) - (*un_data)(l_y_idx)));
+
+        double drag_n = -xi / nu_n * thn_lower * convertToThs(thn_lower) * ((*un_data)(idx) - (*us_data)(idx));
+        double pressure_n = -thn_lower / dx[0] * ((*p_data)(idx_c_up) - (*p_data)(idx_c_low));
+        (*A_un_data)(idx) = D_u * (ddx_Thn_dx_un + ddy_Thn_dy_un + ddy_Thn_dx_vn + ddx_Thn_dy_vn + drag_n) +
+                            D_p * (pressure_n) + C * thn_lower * (*un_data)(idx);
+
+        // solvent equation
+        double ddx_Ths_dx_us = eta_s / (dx[0] * dx[0]) *
+                               (convertToThs((*thn_data)(idx_c_up)) * ((*us_data)(idx + xp) - (*us_data)(idx)) -
+                                convertToThs((*thn_data)(idx_c_low)) * ((*us_data)(idx) - (*us_data)(idx - xp)));
+        double ddy_Ths_dy_us = eta_s / (dx[1] * dx[1]) *
+                               (convertToThs(thn_imhalf_jphalf) * ((*us_data)(idx + yp) - (*us_data)(idx)) -
+                                convertToThs(thn_imhalf_jmhalf) * ((*us_data)(idx) - (*us_data)(idx - yp)));
+        double ddy_Ths_dx_vs = eta_s / (dx[1] * dx[0]) *
+                               (convertToThs(thn_imhalf_jphalf) * ((*us_data)(upper_y_idx) - (*us_data)(u_y_idx)) -
+                                convertToThs(thn_imhalf_jmhalf) * ((*us_data)(lower_y_idx) - (*us_data)(l_y_idx)));
+        double ddx_Ths_dy_vs =
+            -eta_s / (dx[0] * dx[1]) *
+            (convertToThs((*thn_data)(idx_c_up)) * ((*us_data)(upper_y_idx) - (*us_data)(lower_y_idx)) -
+             convertToThs((*thn_data)(idx_c_low)) * ((*us_data)(u_y_idx) - (*us_data)(l_y_idx)));
+        double drag_s = -xi / nu_s * thn_lower * convertToThs(thn_lower) * ((*us_data)(idx) - (*un_data)(idx));
+        double pressure_s = -convertToThs(thn_lower) / dx[0] * ((*p_data)(idx_c_up) - (*p_data)(idx_c_low));
+        (*A_us_data)(idx) = D_u * (ddx_Ths_dx_us + ddy_Ths_dy_us + ddy_Ths_dx_vs + ddx_Ths_dy_vs + drag_s) +
+                            D_p * (pressure_s) + C * convertToThs(thn_lower) * (*us_data)(idx);
```

Comment:
I considered a lot of different ways to refactor this. It would be great to make this modular in some way. You could register a function that specifies how to, e.g., compute the pressure gradient or the drag term, or the viscous force.

The reason I didn't do that is because these are the numerically expensive routines. I want the compiler to see as much of the code as it can so that it can generate efficient code. I left it as this because I didn't want (nor do I have time) to test efficiencies. Ideally, this should be replaced with Fortran code anyway.